--- telepathy-gabble-0.9.13/src/connection.c~	2010-06-14 15:57:03.000000000 +0200
+++ telepathy-gabble-0.9.13/src/connection.c	2010-06-18 15:44:05.279021780 +0200
@@ -1690,24 +1690,25 @@
   if (disco_error != NULL)
     {
       DEBUG ("Got disco error on bare jid: %s", disco_error->message);
-      return;
     }
-
-  for (i = node_iter (result); i; i = node_iter_next (i))
+  else
     {
-      LmMessageNode *child = node_iter_data (i);
-
-      if (!tp_strdiff (child->name, "identity"))
+      for (i = node_iter (result); i; i = node_iter_next (i))
         {
-          const gchar *category = lm_message_node_get_attribute (child,
-              "category");
-          const gchar *type = lm_message_node_get_attribute (child, "type");
-
-          if (!tp_strdiff (category, "pubsub") &&
-              !tp_strdiff (type, "pep"))
+          LmMessageNode *child = node_iter_data (i);
+    
+          if (!tp_strdiff (child->name, "identity"))
             {
-              DEBUG ("Server advertises PEP support in our jid features");
-              conn->features |= GABBLE_CONNECTION_FEATURES_PEP;
+              const gchar *category = lm_message_node_get_attribute (child,
+                  "category");
+              const gchar *type = lm_message_node_get_attribute (child, "type");
+    
+              if (!tp_strdiff (category, "pubsub") &&
+                  !tp_strdiff (type, "pep"))
+                {
+                  DEBUG ("Server advertises PEP support in our jid features");
+                  conn->features |= GABBLE_CONNECTION_FEATURES_PEP;
+                }
             }
         }
     }
